### ISSUE TITLE
Upwind navigation returns wrong results

### ISSUE THREAD

**Issue by dakk** [dakk/libweatherrouting#7]
As reported by @enricofer :

> Start at 09:48 del 22/10/2021 from E25.27515,N35.80502 to E24.73033,N35.81287  (E to W)

[test02.grb.tar.gz](https://github.com/dakk/libweatherrouting/files/7400292/test02.grb.tar.gz)

It returns a wrong result:

![image](https://user-images.githubusercontent.com/1060425/138496625-f37467b0-6b28-4d35-8e4b-b8a7e761e76e.png)

We need to create a test case to check if libwr behavior is correct

**Comment by enricofer** [comment]
I'm having trouble checking results for upwind conditions.
I'm modified the test suite to check upwind navigation for all directions: https://github.com/enricofer/libweatherrouting/blob/9a876316a0030e2b631f6458a2f2924aab919d44/tests/linearbestisorouter_test.py#L69 Basically, The test case get a track ciclying from a base point to all quadrants, mocking grib with a constant upwind in reverse track direction.
For all directions I'm getting a reasonable result but for upwind navigation in north direction I'm getting this not expected result: https://gist.github.com/enricofer/f064c587e13588ef660a4ecf89577d8b
These are all test output together in one geojson: https://gist.github.com/enricofer/411843c5d1cd9e73292d84ebf9880358
I don't understand why.

**Comment by enricofer** [comment]
Furthermore, is it right this result?
I included a [test on routage speed](https://github.com/enricofer/libweatherrouting/blob/9a876316a0030e2b631f6458a2f2924aab919d44/tests/polar_test.py#L36), I would expect zero for twa=0, as for getSpeed method, but I get different result....
```
Traceback (most recent call last):
File "/home/travis/build/enricofer/libweatherrouting/tests/polar_test.py", line 36, in test_routage
self.assertEqual(self.polar_obj.getRoutageSpeed(10,math.radians(0)), 0, 0.001)
AssertionError: 4.443057770090072 != 0 : 0.001
```

**Comment by dakk** [comment]
Sorry for the delay; I tried your tests and results are wrong as you pointed out.
Atm I've no time for a deep debug, but try this: here https://github.com/enricofer/libweatherrouting/blob/9a876316a0030e2b631f6458a2f2924aab919d44/weatherrouting/routers/router.py#L88 instead of using getRoutageSpeed, use getSpeed; your test seems to give a correct result:

```geojson
{"type": "FeatureCollection", "features": [{"type": "Feature", "id": 1, "geometry": {"type": "Point", "coordinates": [17.0, 34.5]}, "properties": {"strait": "mean"}}, {"type": "Feature", "id": 1, "geometry": {"type": "LineString", "coordinates": [[17.05, 34.5], [18.0, 35]]}, "properties": {"strait": "north"}}, {"type": "Feature", "id": 1, "geometry": {"type": "LineString", "coordinates": [[16.95, 34.5], [16.0, 34]]}, "properties": {"strait": "south"}}, {"type": "Feature", "id": 0, "geometry": {"type": "Point", "coordinates": [17, 34]}, "properties": {"timestamp": "2021-04-02 12:00:00", "attr1": 0, "attr2": 0, "attr3": 0, "attr4": 0}}, {"type": "Feature", "id": 1, "geometry": {"type": "Point", "coordinates": [16.887493858644547, 34.13356039140768]}, "properties": {"timestamp": "2021-04-02 13:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 5.275, "attr4": 325.0}}, {"type": "Feature", "id": 2, "geometry": {"type": "Point", "coordinates": [17.050774167231765, 34.269009846086995]}, "properties": {"timestamp": "2021-04-02 14:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 6.2, "attr4": 45.0}}, {"type": "Feature", "id": 3, "geometry": {"type": "Point", "coordinates": [16.93790920922015, 34.40256384863036]}, "properties": {"timestamp": "2021-04-02 15:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 5.275, "attr4": 325.0}}, {"type": "Feature", "id": 4, "geometry": {"type": "Point", "coordinates": [17.101712917701974, 34.53800624320647]}, "properties": {"timestamp": "2021-04-02 16:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 6.2, "attr4": 45.0}}, {"type": "Feature", "id": 5, "geometry": {"type": "Point", "coordinates": [16.962186228592657, 34.67529960820302]}, "properties": {"timestamp": "2021-04-02 17:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 5.8, "attr4": 320.0}}, {"type": "Feature", "id": 6, "geometry": {"type": "Point", "coordinates": [17.126527752866963, 34.81073481562158]}, "properties": {"timestamp": "2021-04-02 18:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 6.2, "attr4": 45.0}}, {"type": "Feature", "id": 7, "geometry": {"type": "Point", "coordinates": [16.98654060783366, 34.948021207463654]}, "properties": {"timestamp": "2021-04-02 19:00:00", "attr1": 0.0, "attr2": 10.0, "attr3": 5.8, "attr4": 320.0}}, {"type": "Feature", "id": 8, "geometry": {"type": "LineString", "coordinates": [[17, 34], [16.887493858644547, 34.13356039140768], [17.050774167231765, 34.269009846086995], [16.93790920922015, 34.40256384863036], [17.101712917701974, 34.53800624320647], [16.962186228592657, 34.67529960820302], [17.126527752866963, 34.81073481562158], [16.98654060783366, 34.948021207463654]]}, "properties": {"start-timestamp": "2021-04-02 12:00:00", "end-timestamp": "2021-04-02 19:00:00"}}]}
```

![image](https://user-images.githubusercontent.com/1060425/144410756-fa7234f2-f51d-4f0a-b934-ba6dfce73251.png)


I think using getSpeed instead of getRoutageSpeed is correct, since creating isochrones we need the effective speed retrieved from polar (or an interpolated value for missing angles). 
getRoutageSpeed instead, gives a value I'm not quite sure what it is.

Please give it a try and let me know

**Comment by enricofer** [comment]
You are right. using getSpeed i'm getting much more reasonable results: https://gist.github.com/enricofer/e053c9ce45fb20d7cbafd595b17ab8de
The getRoutageSpeed method as far as I understand calculate the best VGM https://it.wikipedia.org/wiki/Velocity_Made_Good for a gived wind relative angle, but iterating over ALL direction to find the most effective route is a concept that overcomes and contains the search for the the best VMG. 
Comparing the test results for getSpeed evaluation method with those for getRoutageSpeed evaluation method: https://gist.github.com/enricofer/411843c5d1cd9e73292d84ebf9880358, keeping apart the faulty 000 heading, appears that the methods outputs a longest route but allowing faster speed than the second method but the reaching time is the same, and therefore, since wind propellent is free, the result is the same, but increasing algorithm performances.
For completeness, tests should be performed in a more real conditions but I'm expecting the same result so I would get rid of getRoutageSpeed method for best route evaluation. If you agree, I would submit a PR.
ping @pcav

**Comment by pcav** [comment]
Fine to me, thanks for letting me know @enricofer

**Comment by dakk** [comment]
@enricofer agree for the PR, getSpeed is the correct approach for isochrones calculations: the VMG is intrinsic here, since the isochrones tree will return fastest route anyway.
We can create another router class using VMG calculations without isochrones.

Waiting for PR.

**Comment by dakk** [comment]
Closed by PR #9
